packages/react/src/React.js

```diff
@@ -48,57 +48,109 @@ import {
 import ReactSharedInternals from './ReactSharedInternals';
 import {enableStableConcurrentModeAPIs} from 'shared/ReactFeatureFlags';
 
-const React = {
-  Children: {
-    map,
-    forEach,
-    count,
-    toArray,
-    only,
-  },
-
-  createRef,
-  Component,
-  PureComponent,
-
-  createContext,
-  forwardRef,
-  lazy,
-  memo,
-
-  Fragment: REACT_FRAGMENT_TYPE,
-  StrictMode: REACT_STRICT_MODE_TYPE,
-  Suspense: REACT_SUSPENSE_TYPE,
-
-  createElement: __DEV__ ? createElementWithValidation : createElement,
-  cloneElement: __DEV__ ? cloneElementWithValidation : cloneElement,
-  createFactory: __DEV__ ? createFactoryWithValidation : createFactory,
-  isValidElement: isValidElement,
-
-  version: ReactVersion,
-
-  __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED: ReactSharedInternals,
-};
-
-if (enableStableConcurrentModeAPIs) {
-  React.ConcurrentMode = REACT_CONCURRENT_MODE_TYPE;
-  React.Profiler = REACT_PROFILER_TYPE;
+// Please make sure that no properties are added to this object after its
+// creation. This ensures the object keeps the same shape for performance reasons.
+let React;
+// We should remove the need for two objects if hooks are enabled by default.
+if (enableHooks) {
+  React = {
+    Children: {
+      map,
+      forEach,
+      count,
+      toArray,
+      only,
+    },
+
+    createRef,
+    Component,
+    PureComponent,
+
+    createContext,
+    forwardRef,
+    lazy,
+    memo,
+
+    Fragment: REACT_FRAGMENT_TYPE,
+    StrictMode: REACT_STRICT_MODE_TYPE,
+    Suspense: REACT_SUSPENSE_TYPE,
+
+    createElement: __DEV__ ? createElementWithValidation : createElement,
+    cloneElement: __DEV__ ? cloneElementWithValidation : cloneElement,
+    createFactory: __DEV__ ? createFactoryWithValidation : createFactory,
+    isValidElement: isValidElement,
+
+    version: ReactVersion,
+
+    __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED: ReactSharedInternals,
+
+    ConcurrentMode: enableStableConcurrentModeAPIs
+      ? REACT_CONCURRENT_MODE_TYPE
+      : null,
+    Profiler: enableStableConcurrentModeAPIs ? REACT_PROFILER_TYPE : null,
+
+    unstable_ConcurrentMode: !enableStableConcurrentModeAPIs
+      ? REACT_CONCURRENT_MODE_TYPE
+      : null,
+    unstable_Profiler: !enableStableConcurrentModeAPIs
+      ? REACT_PROFILER_TYPE
+      : null,
+
+    useCallback,
+    useContext,
+    useEffect,
+    useImperativeMethods,
+    useLayoutEffect,
+    useMemo,
+    useMutationEffect,
+    useReducer,
+    useRef,
+    useState,
```

Comment:
Curious why we forked for `enableHooks` rather than using ternaries like for `enableStableConcurrentModeAPIs` and `__DEV__`? e.g.
```js
React = {
  Children: {
    map,
    forEach,
    count,
    toArray,
    only,
  },

  createRef,
  Component,
  PureComponent,

  createContext,
  forwardRef,
  lazy,
  memo,

  Fragment: REACT_FRAGMENT_TYPE,
  StrictMode: REACT_STRICT_MODE_TYPE,
  Suspense: REACT_SUSPENSE_TYPE,

  createElement: __DEV__ ? createElementWithValidation : createElement,
  cloneElement: __DEV__ ? cloneElementWithValidation : cloneElement,
  createFactory: __DEV__ ? createFactoryWithValidation : createFactory,
  isValidElement: isValidElement,

  version: ReactVersion,

  __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED: ReactSharedInternals,

  ConcurrentMode: enableStableConcurrentModeAPIs
    ? REACT_CONCURRENT_MODE_TYPE
    : null,
  Profiler: enableStableConcurrentModeAPIs ? REACT_PROFILER_TYPE : null,

  unstable_ConcurrentMode: !enableStableConcurrentModeAPIs
    ? REACT_CONCURRENT_MODE_TYPE
    : null,
  unstable_Profiler: !enableStableConcurrentModeAPIs
    ? REACT_PROFILER_TYPE
    : null,

  useCallback enableHooks ? useCallback : null,
  useContext enableHooks ? useContext : null,
  useEffect enableHooks ? useEffect : null,
  useImperativeMethods enableHooks ? useImperativeMethods : null,
  useLayoutEffect enableHooks ? useLayoutEffect : null,
  useMemo enableHooks ? useMemo : null,
  useMutationEffect enableHooks ? useMutationEffect : null,
  useReducer enableHooks ? useReducer : null,
  useRef enableHooks ? useRef : null,
  useState enableHooks ? useState : null,
};
```
No big deal I guess, just curious.

Reply:
I did that originally, but @gaearon suggested that it unnecessarily exposed the hooks APIs as properties.

Reply:
I had to draw the line somewhere and they seemed less bad to expose. Forking the object 4 times would have been pretty horrible. :D

Reply:
Yeah I agree completely. I wasn't suggesting that 😅 I don't like even this fork but I think we've already talked about it more than it's worth so I'll 🤐

Reply:
Hmm to be clear I don't think it's great to expose `ConcurrentMode` = null either.

Reply:
Would you feel any differently about `undefined`?